zlib-rs/src/cpu_features.rs

```diff
@@ -27,10 +27,26 @@ pub fn is_enabled_sse42() -> bool {
 }
 
 #[inline(always)]
-pub fn is_enabled_avx2() -> bool {
+pub fn is_enabled_avx2_and_bmi2() -> bool {
     #[cfg(any(target_arch = "x86_64", target_arch = "x86"))]
     #[cfg(feature = "std")]
-    return std::is_x86_feature_detected!("avx2");
+    {
+        use std::sync::atomic::{AtomicU32, Ordering};
+
+        static CACHE: AtomicU32 = AtomicU32::new(2);
+
+        return match CACHE.load(Ordering::Relaxed) {
+            0 => false,
+            1 => true,
+            _ => {
+                let detected = std::is_x86_feature_detected!("avx2")
+                    && std::is_x86_feature_detected!("bmi1")
+                    && std::is_x86_feature_detected!("bmi2");
+                CACHE.store(u32::from(detected), Ordering::Relaxed);
+                detected
+            }
+        };
+    }
```

Comment:
because https://github.com/rust-lang/libs-team/issues/585

Reply:
You could avoid the cache write if all these target features are statically enabled.